### PR TITLE
Proposed spec for backend (DRAFT)

### PR DESCRIPTION
**Pull request by rtw43**
a big ugly markdown file that really needs:
A) spellchecking
B) splitting into smaller ones (for stuff in forewords esp)
C) an author with more motivation to do the above ;)

but outlines my idea of backend specification

### CODE REVIEW

**Comment by shiyah** [comment]
Small thing:
[Project grants](https://github.com/rtw43/cs3099jh/blob/843330a3ef0e8b503385d11fbce562be7f7a8aaa/protocols/backend_spec.md#project-grants) do not issue an error if the new access level does not exist

**Comment by rtw43** [comment]
@shiyah The intention here was that this would be considered an invalid request (and be dealt with as specified in the basic response section). I have clarified the text in my local copy and will release an updated version soon, under a new version number.

This new version will also provide an endpoint for the client to list all valid access_levels and privileges.

Does this seem appropriate? Would you prefer a customer error along the lines of "invalid_access_level" or "access_level_not_found"?

**Comment by shiyah** [comment]
@rtw43 
In theory, the front-end would query the API for access_levels beforehand and only allow those to be chosen, making an invalid access_level likely only possible in two scenarios:
* access_levels is changed between available access levels query and the project grant query
* Testing of the endpoint to see how it reacts to various invalid requests

In both cases, an `invalid_request` response would not be as clear as to what the error is (did the request not match the pattern or is the access_level wrong?). Splitting `access_level_not_found` off therefore makes a good amount of sense.

Also, thank you for the sheer amount of work you put into making the backend spec.

**Comment by rtw43** [comment]
@shiyah 
This reasoning is fair enough, as part of the 0.1.3 release I have added a specific error for this case (as well as the equivalent case when an invalid privilege is specified).

**Comment by rtw43** [comment]
I have made some additional resources for BE01 available here https://rw86.host.cs.st-andrews.ac.uk/jh/extra_resources.md.html

**Comment by db213** [comment]
Umm... out of curiosity, what was wrong with the ML protocol I had written, and why couldn't I just have fixed whatever issues there were instead of completely rewriting it? (I'm sorry, not trying to be mean and definitely not claiming what I did was perfect, but it really doesn't feel good to have done some work and just to have it ignored and rewritten...)

https://swaggerhub.com/apis/JH-Project/machine-learning-api/1.0

**Comment by rtw43** [comment]
Hi,

I am sorry - I don't wish to be seen as being combative, overreaching or counter productive.

Your existing protocol notes were indeed my source of inspiration (many thanks), and my original intention was to present my work as a development of them directly. However, I found myself approaching the problem from a fundamentally different position than the original notes did, in some ways like a job centric view rather than a model centric view, a different view of the role of backend/hci, such as who executes the calls, who stores/responds with the data, what kinds of tasks ML will perform, if ML is responsible for generating the report, what format the output should take, and several other things that I note under the "Acknowledgements" section.

I don't know if these changes make the design "better" but I do know that they are different to what you intended when writing the notes - and I wish people to contrast them side by side. I really do want to hear about all the reasons these changes are stupid and why "functionality x belongs instead in component y" and why "the jobs model captures the use case x badly".

This lead me to feel that the changes would be better served if presented as a alternative architecture: I no longer felt comfortable proposing such overarching changes to your work (since it seemed to conflict with some of your design intentions when you originally wrote it). My hope is that the relative merits of both approaches will be considered, I am of the belief that the best engineering occurs when several people come up with competing approaches and architectures. And think the year as whole will benefit more from exploring several different architectures than from exploring several revisions of the same one. In particular I want to give you an opportunity to show how "we can achieve the functionality you listed here much better if we instead adapt my design like so".

I am sorry if it comes across as trying to overrule your work, my intention was quite the opposite! To recognise that these are two fundamentally different designs and encourage people to think about them and debate them as such. I do not feel it would have been appropriate to pose my work as a development of yours (and if I were in your position I would have been more offended by the attempt to pass off a fundamental rearchitecture as an "improvement").

I should note: I have been liberal in slapping the word "proposal" all over everything I write, and I am under no disillusions that the year have to or even should go with them - but I did want an opportunity to present a cohesive proposal, that reflected how things fit together in my head. I encourage you to critique as much as possible and hope that I have not caused animosity.

Best,
Ryan <3

**Comment by db213** [comment]
Hi Ryan, thanks for the response. :) I didn't mean to make you feel bad, it was just a bit of jolt to see this when I finally started looking at JH stuff again. Having an alternative is good, I just feel like it's hard to compare when mine was literally only the ML protocol.

I'll start reading this over and leaving comments on it. :)
Edit: sorry I left loads of comments, I really hope I don't come off as harsh, I don't mean to.

File: protocols/backend_spec.md [hunk 1]
```diff
@@ -0,0 +1,1326 @@
...
+{
+    "status": "error",
+    "error": "invalid_request",
+    "error_description": "The client made an invalid request."
+}
+```
+## Internal errors
+At any time the server **MAY** respond to a request with response code 500 and return error "internal_server_error".
+
+# OAuth
```
Comment (db213) [inline protocols/backend_spec.md:367]: Alex already suggested in the slack that OAuth might be overkill. Seems to me like it is, but maybe we should vote on it?

File: protocols/ML01.md [hunk 2]
```diff
@@ -0,0 +1,315 @@
...
+- First draft
+
+# Foreword: Acknowledgements
+
+This specification is loosely based on ideas presented in an endpoint outline "ml_endpoints.md" written by Daphne Bogosian. There are some major architectural differences, outlined below:
+
+- The principal "backend should be responsible for all matters of persistance" agreed during protocol meetings has been incorporated
+  - In particular results are returned to the frontend as files via backend (this was also agreeed)
+- The principal "HCI should talk directly to backend" has been incorporated.
+  - In particular there is no BE <-> ML communication.
```
Comment (db213) [inline protocols/ML01.md:30]: Is this a mistake? I thought we agreed there would be no HCI <-> ML communications, only HCI -> BE -> ML?

File: protocols/ML01.md [hunk 3]
```diff
@@ -0,0 +1,315 @@
+# ML01: Machine learning basic specification
+**THIS IS A DRAFT SPECFICATION AND IS NOT FINAL**
+
+| name                       | ML01               |
+|----------------------------|--------------------|
+| version                    | 0.1.0              |
+| status                     | proposal           |
+| author                     | Ryan Wilson (rw86) |
+| serving component(s)       | Machine Learning   |
+| consuming component(s)     | HCI                |
```
Comment (db213) [inline protocols/ML01.md:10]: But... but.... we voted against this... like the whole class voted??? :'(

File: protocols/ML01.md [hunk 4]
```diff
@@ -0,0 +1,315 @@
...
+- The principal "backend should be responsible for all matters of persistance" agreed during protocol meetings has been incorporated
+  - In particular results are returned to the frontend as files via backend (this was also agreeed)
+- The principal "HCI should talk directly to backend" has been incorporated.
+  - In particular there is no BE <-> ML communication.
+- The API has been generalised to not specifically reference training/models and instead refence the more general "jobs":
+  - Gives more flexibility to the implementation for data processing tasks that do not fit this pattern.
+  - Removes burden on client.
+  - Jobs referencing models are still posibile (see implementation notes)
+- Response format made consistent with BE01.
+- Spec now only uses GET and POST, as these are the request types with most widely available browser support.
```
Comment (db213) [inline protocols/ML01.md:36]: I thought mostly all browsers supported the other request types as well?

File: protocols/ML01.md [hunk 5]
```diff
@@ -0,0 +1,315 @@
...
+
+# Foreword: URL
+
+Within this document the URL `http://ml.endpoint/` will be used to indicate the base URL of the backend server's API endpoint.
+
+# Foreword: Matching notation
+
+The matching notation used in this specification is identical to that used in BE01.
+
+# Overview
```
Comment (db213) [inline protocols/ML01.md:53]: There is no support in here for recommending ML algorithms, which was a requirement from the contact in medicine.

File: protocols/ML01.md [hunk 6]
```diff
@@ -0,0 +1,315 @@
...
+http://ml.endpoint/projects/<project_name>/job_types
+```
+
+If the given project does not exist the server **SHOULD** return error code 404 and error "project_not_found".
+
+If the user does not have accessed to the project an unauthorised response **SHOULD** be returned. (see above)
+
+If the project exists and the user is authorised the server's unwrapped response **MUST** match
+
+```javascript
```
Comment (db213) [inline protocols/ML01.md:85]: These responses/requests don't follow the JSON schema, which we had previously agreed to use.

File: protocols/ML01.md [hunk 7]
```diff
@@ -0,0 +1,315 @@
...
+
+If the user does not have accessed to the project an unauthorised response **SHOULD** be returned.
+
+If the given job does not exist then the server **SHOULD** return error code 404 and error "job_not_found".
+
+otherwise the server **MUST** return a successful empty response.
+
+# Notes for implementors
+## State
+This specification and BE01 were together written to allow implementations that did not depend on the ML endpoints requiring any storage. As an example of how this may work:
```
Comment (db213) [inline protocols/ML01.md:295]: It's probably just me being an idiot, but I didn't understand this bit. Could you ELI5?

**Review by vv20 — APPROVED** [review]

**Comment by db213** [comment]
This should not have been accepted without the comments being resolved...?
Edit: I'm an idiot. approving != merging

File: protocols/backend_spec.md (hunk 1 above)
Comment (rtw43) [reply protocols/backend_spec.md:367]: Yeah I agree that OAuth is a *HUGE* specification to bring in - In the spec I striped out as much of the OAuth stuff as possible (all stuff about client tokens is gone for instance), and only supported authentication using username and passwords (and refresh tokens) - I feel like this is pretty minimal and a a custom protocol would end up looking pretty similar. Would be interested in Alex's thoughts on this. But yeah a vote to get overall thoughts would be good.

**Comment by CodingCellist** [comment]
@rtw43 Could you respond to the comments please? :-)

Also, ML will need data from BE. How do you intend to do that when ML and BE should not communicate according to your proposal?

File: protocols/backend_spec.md (hunk 1 above)
Comment (db213) [reply protocols/backend_spec.md:367]: Alex had suggest JSON Web Tokens. I honestly don't know much about any of this, so probably better to just contact him directly.

**Comment by rtw43** [comment]
Literally in the processes of doing so as you typed that

File: protocols/ML01.md (hunk 2 above)
Comment (rtw43) [reply protocols/ML01.md:30]: Checking back against the minutes it seems so. Thinking about how I would change it to be HCI -> BE -> ML I think (at least for this design) it makes more sense to me to "cut out the middle man" although, I didn't realise that this was deviating behaviour - mea culpa.

The HCI -> BE -> ML version would, basically involve BE providing an endpoint and forwarding requests and responses - so the protocol wouldn't change except all the URLs looking like "backend.endpoint/ML/...". I still do feel that BE kind of has a lot to do as it is - so to me adding any more intelligence to backend rather than "forwarding the requests" would feel like poor separation of concerns.

File: protocols/ML01.md (hunk 3 above)
Comment (rtw43) [reply protocols/ML01.md:10]: My fault - see above on how this would be changed

File: protocols/ML01.md (hunk 4 above)
Comment (rtw43) [reply protocols/ML01.md:36]: Was talking with Jordan about this as well  - I'm still not entirely clear about the client support for RESTFUL methods but this may have been too conservative a design choice - no issue with changing it.

File: protocols/ML01.md (hunk 2 above)
Comment (db213) [reply protocols/ML01.md:30]: I thought the idea of the HCI -> BE -> ML structure was that the calls made for ML (to BE from HCI) were pretty much just forwarded through to ML (after checking tokens), so shouldn't add to much work for BE. It does make BE a kind of useless middle man in that particular situation, but that was discussed and people still went with the idea, as it greatly simplifies the structure, so I think we should stick with it? (Not sure what you mean by adding more intelligence to BE, cause my understanding was they pretty much just were forwarding requests)?

File: protocols/ML01.md (hunk 4 above)
Comment (db213) [reply protocols/ML01.md:36]: Hmmm, but if you do change this so that requests are just forwarded to ML, there should be no issue making ML requests other than GET/POST since it won't be through a browser?

File: protocols/ML01.md (hunk 6 above)
Comment (rtw43) [reply protocols/ML01.md:85]: assuming this means jsonapi - maybe I misunderstand this specification but it seems to be that the other protocols don't either? Responses "data" attributes don't seem to be valid  “Resource objects”  as the spec requires? A confusion about exactly how and what features of jsonapi are being used  led me to develop the simplest response format that I felt met the needs. I don't foresee massive difficulty in re-mapping the response attributes to match it pending clarification of exactly what features are to be used - personally jsonapi seems a bit overbearing.

File: protocols/ML01.md (hunk 7 above)
Comment (rtw43) [reply protocols/ML01.md:295]: Idea is that ML nodes **could** store all working data in backend without need for any local storage / databases.  This is nothing new - I believe it was you who originally proposed the idea of backend storing MLs models, this is just a restatement of that.

File: protocols/ML01.md (hunk 6 above)
Comment (db213) [reply protocols/ML01.md:85]: I don't really have an opinion on whether the jsonapi is a good idea or not, it's just something that we'd agreed to use. From http://jsonapi.org/format/ the "data" attribute is described:

"
Top Level

A JSON object MUST be at the root of every JSON API request and response containing data. This object defines a document’s “top level”.

A document MUST contain at least one of the following top-level members:

    data: the document’s “primary data”
    errors: an array of error objects
    meta: a meta object that contains non-standard meta-information.
"

I was trying to follow the schema when I wrote the Swagger, but it's certainly possible I made mistakes.

**Comment by rtw43** [comment]
@teh6 

    Also, ML will need data from BE. How do you intend to do that when ML and BE should not communicate according to your proposal?

I muck up of wordings on my part. My intention is that ML acts as a normal BE client (using the standard file APIs that HCI uses), what I meant was that there is no special protocol only spoken between BE and HCI.

File: protocols/ML01.md (hunk 7 above)
Comment (db213) [reply protocols/ML01.md:295]: Ah okay, I was confused because earlier you'd said there is no need BE <-> ML communication, when this would require ML making calls to BE, no?

File: protocols/backend_spec.md (hunk 1 above)
Comment (rtw43) [reply protocols/backend_spec.md:367]: Hrm, In my understand JSON Web Tokens is a format of token and OAuth is way of getting tokens (without specifying what these look like) and in this way they are orthogonal.

File: protocols/ML01.md (hunk 2 above)
Comment (rtw43) [reply protocols/ML01.md:30]: By intelligence I simply meant anything but forwarding. We seem to be on the same page on this. Except I am not too sure how ML and BE would authenticate the requests to each in this case? would ML still use the users token?

File: protocols/ML01.md (hunk 4 above)
Comment (rtw43) [reply protocols/ML01.md:36]: But the browser would still make a request to BE and (unless the BE modifies the request method before forwarding) there is still a question of client support.

File: protocols/ML01.md (hunk 2 above)
Comment (db213) [reply protocols/ML01.md:30]: Well I had this idea that BE would check that the user had permissions for the files HCI asked to ML. So if they didn't, the request would never even reach ML, just get an error code sent back. That way ML doesn't need to worry about tokens - if anything ever reaches them, they know there should be no issues with permissions so would not need to use user's tokens.

File: protocols/ML01.md (hunk 4 above)
Comment (db213) [reply protocols/ML01.md:36]: Yeah, you're right. I think most browsers should support other requests though, and it would make the API a lot nicer to use some of the other ones.

File: protocols/ML01.md (hunk 5 above)
Comment (rtw43) [reply protocols/ML01.md:53]: I saw this as just being a special job type offered by ML - just one that does not have a specific algorithm, whereby ML produces a report indicating possible correlations / suggested specific algorithms. I think you envision more of a way of "filtering" the algorithm list? This would be as simple as adding optional parameter(s) to the endpoint that lists job types. If this is what you're referring to then this seems a perfectly straightforward and clean change to me.

**Comment by db213** [comment]
So I'm realising that I misunderstood quite a bit of this - with changes to make it fit the structure of HCI -> BE -> ML, I'm wondering if it would actually be possible to still use (possibly an edited version of) what I came up with? I don't actually see anything too drastically different to make that impossible and I would feel a lot better to know it might still get some use/not have been wasted effort.

That being said, I didn't understand your concept of "jobs", and how they differ from the model-centric idea I used, so maybe that's where the difference lies?

File: protocols/ML01.md (hunk 4 above)
Comment (rtw43) [reply protocols/ML01.md:36]: Jordan's issue: https://github.com/rtw43/cs3099jh/issues/4 (this was specifically backend, but relevant)

File: protocols/ML01.md (hunk 2 above)
Comment (db213) [reply protocols/ML01.md:30]: It also just occurred to me that BE would have to do a little bit more work when the user asks to train a model in that they will need to first check if that model exists already or not. But I think this would need to be done somehow no matter what?

File: protocols/ML01.md (hunk 6 above)
Comment (rtw43) [reply protocols/ML01.md:85]: I was referring to 
```
A document MUST contain at least one of the following top-level members:
...
data: the document’s “primary data”
```

```
Primary data MUST be either:

    a single resource object, a single resource identifier object, or null, for requests that target single resources
    an array of resource objects, an array of resource identifier objects, or an empty array ([]), for requests that target resource collections
```
and
```
A resource object MUST contain at least the following top-level members:

    id
    type
```

**Comment by rtw43** [comment]
Unfortunately, due to increasing time pressure and other priorities in my life, I will no longer be seeking to merge my proposals, and will be withdrawing my pull request imminently. While my work was originally submitted with the best of intentions, I understand it has not been wholly accepted as such. My work will still be accessible via the original repo, so it may remain as a resource for anybody who wishes to utilize it. I continue to believe that the current protocols have significant issues, both architecturally and in terms of clarity and assumptions made, which I have already devoted some weeks to attempting to convey. Whilst I am no longer able to devote such large amounts of time to this, I urge continued revision and increased documentation so that these issues may both be made evident and hopefully resolved. Shortly, I will update my proposals to give explicit permission for editing and derivative works (subject to some minor conditions). I am open to anybody building their own proposal on the basis of mine, or if they possess an opinion on the matter, resubmitting my proposal; but I do not feel I currently have the time (or the popular backing) to devote the necessary attention to defending my pull request.

Without further motive, I would like to comment on my personal disagreement with the current governance structure of the project - wherein major decisions are made in private channels and discussions with unelected members, rather than in a public forum. I think we can all agree it would be best to debate publicly in order to give everyone the opportunity to engage and input, and in the hopes of encouraging merit based discussion that will ultimately result in the best project for all of us.

With thanks for your time and engagement, 
    - Ryan (with Mhairi on spellcheck)

**Comment by theoden8** [comment]
I do think that Ryan's spec is a good attempt to improve the existing project. I myself do not feel I have the competency to standardize or assess a viable protocol, being a Joint Honours student. However, I would suggest that we move the discussion into the meeting rather than have an argument amongst the few people who are a very limited representation of students on the module, and therefore could be prone to bias towards personal credit.

I therefore ask to hold the withdrawal until we have a meeting, which, hopefully, will be organized shortly.

Best,
Kirill

**Comment by CodingCellist** [comment]
@rtw43
Dear Ryan,  

Firstly, we would like to apologise if Thomas’ e-mail caused you any offense. There seems to have been a misunderstanding in terms of what we wanted to achieve by asking you to stick to the agreed protocols which we believed had been agreed through an open discussion involving all project participants. Below we have tried to outline how to move forward by merging the work done by you with the work done by others and the structure jointly agreed by all participants.  

Our starting point has always been that it is best to discuss things publicly. This is why we initially held a StrawPoll on the preferred format for such discussions, the conclusion of which was that a number of meetings open to all should be held. Three such meetings took place throughout the first semester, and the current structure was agreed by the approx. 20 participants at the final one: Namely that HCI would communicate only with BE and that ML would communicate only with BE (i.e. the HCI <=> BE <=> ML structure). Any reservations with this approach could have been raised either in that meeting or following the subsequent distribution of the minutes.  

However, the considerable work you have done on the protocols is not incompatible with the approach above, and we by no means want to discard what you have done. On the contrary, we think it could be useful to merge it with the existing work done by Daphne as obviously otherwise there will be no APIs or protocols for HCI or BE. However, the overall structure must remain the one that was decided by all in the final meeting before the holidays.  

If there are issues with the current protocols – and we share your point that there is – we do indeed need to address them. However, it is important that such issues are raised on Git, i.e. on the official repo and not a fork, since Git allows people to know that there is an issue, to discuss alternatives, to put themselves down as developers, to easily track what is still open and what has been resolved and by whom, etc.  

We hope that the above is acceptable to you and look forward to continue working together on this complex project.  

All the best,
Thomas Hansen, Daphne Bogosian, and Aidan McGinley

**Comment by Knepsia** [comment]
As I see it, there was no protocol here for Back End or Front End, and Ryan provided one for those two, he also further defined the existing Machine Learning protocol, explaining it more clearly and making the changes as was agreed in the meetings. While I agree it should all be kept in the repository where all the protocols are created, Ryan did so, he made a pull request 28 days ago...
So what was the issue here?
